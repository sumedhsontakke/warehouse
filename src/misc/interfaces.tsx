import { FormEvent, ReactNode } from "react";

export interface IProductsContext{
    products: IProduct[] | null;
    setProducts: (products: IProduct[]) => void    
    saveProductList: (retry: number) => void
}

export interface IProduct{
    id: string;
    name: string;
    requiredArticles: IRequiredArticle[]
} 

export interface IProductSave {
    name: string;
    articles: IRequiredArticleState[]    
}

export interface IRequiredArticle{
    id: string;
    amountRequired: number;
}

export interface IRequiredArticleState{
    id: string;
    amountRequired: number | string;
}

export interface IArticle{
    id: string;
    name: string;
    amountInStock?: number;  
    amountToSubtract?: number  
}


export interface IArticleDetails extends IRequiredArticle, IArticle{

}

export interface IProductDetailsParam{
    productId: string
}

export interface IProductDetailsWithArticles extends IProduct {
    articles: IArticleDetails[]
}

export interface IProductDetailsProps extends IProductDetailsWithArticles{
    saleProduct: (id:string, articles:IArticleDetails[]) => void;
}

export interface IReactNode{
    children: ReactNode;
}

export interface ICustomButton extends IReactNode {
    type?: string;
    onClick?: (e:Event) => any;
    margin? : string;
    width?: string;
    disabled?: boolean | null | undefined;
}

export interface IText extends IReactNode{
    fontSize?: string;
    fontWeight?: number; 
    color? : string;
    margin? : string;
    onClick? : () => void;
}

export interface IFlexWrap {
    flexDirection?: string;
    children?: ReactNode;
}

export interface IModalProps{
    showModal: boolean;
    setShowModal: () => void;
    handleClose: () => void;
    handleSubmit: () => void;
    headingText?: string;
    bodyText?: string; 
    closeButtonText?: string; 
    submitButtonText?: string; 
}

export interface ISale {
    amountSold: number;
    createdAt: string;
    id: string;
    productId: string;
    name: string;
}

export interface ISalesProps {
    salesDetail: ISale[] | null;
    isLoading: boolean;
    productList: IProduct | null;
}

export interface IArticleDropdrown extends IArticle {

}

export interface ISetProduct {
    name: string;
    requiredArticles: IRequiredArticle[];
}

export interface IAlert{
    message: string;
    variant: string;
    show: boolean;
}

export interface IAddProductProps {
    submit : (e: FormEvent) => void;
    setProduct: (value: React.ChangeEvent<HTMLInputElement>) => void;
    addRequiredArticle: (value: React.ChangeEvent<HTMLInputElement | HTMLSelectElement> | null, inputType: string, index: number) => void;
    allArticles: IArticle[] | null;
    requiredArticles: IRequiredArticleState[];
    alert: IAlert;
    setAlert: (alert: IAlert) => void;
    product: string;
}

export interface IArticlesProps {
    articles: IArticleState[] | null;
    updateArticle: (id: number) => void;
    setArticles: (article: IArticleState[]) => void;
}

export interface IArticleState extends IArticle{
    inputValue: string | number;
}

export interface IArticleComponentProps{
    article: IArticleState;
    updateArticle: (id: number) => void; 
    setArticles: (value: string, index: number) => void;
    articleIndex: number;
}

export interface IUpdateArticle {
    name: string;
    amountInStock: number
}

