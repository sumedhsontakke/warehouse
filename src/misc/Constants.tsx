export const ROOT_URL = "http://localhost:7000/";
export const PRODUCTS_URL = "products/";
export const ARTICLES_URL = "articles/";
export const SALES_URL = "sales/";