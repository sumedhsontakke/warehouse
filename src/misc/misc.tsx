import { IRequiredArticleState, IArticle } from "./interfaces";
export const validate = (
    product: string,
    requiredArticles: IRequiredArticleState[],
    allArticles: IArticle[]
): string => {

    let validateResult = ""
    requiredArticles.map(article => {
        if (article.id === "") {
            validateResult = "Select article";
        } else if (article.amountRequired === 0) {
            validateResult = "Enter article amount";
        }
        const filteredArticle: IArticle[] = allArticles.filter((_article: IArticle) => _article.id === article.id);
        console.log(filteredArticle);
        if (
            filteredArticle.length > 0 &&
            filteredArticle[0].amountInStock &&
            article.amountRequired &&
            filteredArticle[0].amountInStock < article.amountRequired
        ) {
            validateResult = "Entered article quantity not available, please add article";
        }
    });
    if (product == "") {
        validateResult = "Enter product name";
    }

    const uniqueIds = new Set(requiredArticles.map(article => article.id));
    if (uniqueIds.size < requiredArticles.length) {
        validateResult = "Select unique Articles";
    }

    return validateResult;
}