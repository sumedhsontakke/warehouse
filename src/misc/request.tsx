import { ROOT_URL, PRODUCTS_URL, ARTICLES_URL, SALES_URL } from "./Constants";
import axios from 'axios';
import { IArticle, IProductSave, IUpdateArticle } from "./interfaces";

export const getProductList = () => {
    return axios.get(`${ROOT_URL}${PRODUCTS_URL}`)
}

export const getProductDetails = (id: string) => {
    return axios.get(`${ROOT_URL}${PRODUCTS_URL}${id}`)
}

export const getArticleDetails = (id?: string) => {
    if (id) {
        return axios.get(`${ROOT_URL}${ARTICLES_URL}${id}`)
    } else {
        return axios.get(`${ROOT_URL}${ARTICLES_URL}`)
    }
}

export const getAllArticles = async () => {
    return await axios.get(`${ROOT_URL}${ARTICLES_URL}`)
        .then(resp => resp)
        .catch(err => err)
}

export const updateArticleById = async (id: string, article: IUpdateArticle) => {
    return await axios.patch(`${ROOT_URL}${ARTICLES_URL}${id}`, article,
        {
            headers: {
                'Content-type': 'application/json; charset=UTF-8',
                "Access-Control-Allow-Origin": "*"
            }
        })
        .then(resp => resp)
        .catch(err => err)
}

export const salesProduct = (productDetails: any) => {
    return axios.post(`${ROOT_URL}${SALES_URL}`, { ...productDetails })
}

export const saleArticles = (articles: IArticle[]) => {
    return axios.patch(`${ROOT_URL}${ARTICLES_URL}`, articles,
        {
            headers: {
                'Content-type': 'application/json; charset=UTF-8',
                "Access-Control-Allow-Origin": "*"
            }
        }
    );
}

export const getSales = () => {
    return axios.get(`${ROOT_URL}${SALES_URL}`)
}

export const saveProduct = async (product: IProductSave) => {
    return await axios.post(`${ROOT_URL}${PRODUCTS_URL}`, product)
        .then(resp => resp)
        .catch(err => err)
}
