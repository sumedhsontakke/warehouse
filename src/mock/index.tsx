import { AxiosResponse } from "axios";
import { IProduct } from "../misc/interfaces";
export const product: IProduct[] = [
    {
      "id": "a269a247-0d38-4b47-9630-79c9ae545b68",
      "name": "Dining Chair",
      "requiredArticles": [
        {
          "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
          "amountRequired": 4
        },
        {
          "id": "831b92b8-677b-42cc-a585-335ea4ccccb6",
          "amountRequired": 1
        },
        {
          "id": "addc65a8-c759-41d8-a18a-89fe446ad484",
          "amountRequired": 8
        }
      ]
    }
  ];


export const mockresponse: AxiosResponse = {headers: {}, config: {}, statusText: "", status: 100, data: product};

export const salesMockResponse: AxiosResponse = {headers: {}, config: {}, statusText: "", status: 100,
     data:[
        {
          "id": "da5846e7-ba53-4189-b062-81fc64582431",
          "createdAt": "2020-11-13T12:00:00.123",
          "productId": "a269a247-0d38-4b47-9630-79c9ae545b68",
          "amountSold": 1
        }
     ] 
};

export const articleMockResponse: AxiosResponse = {headers: {}, config: {}, statusText: "", status: 100,
    data:[
        {
            "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
            "name": "Leg",
            "amountInStock": 12
        },
    ] 
};

export const  SalesProps = {
    salesDetail: [
        {
            "id": "da5846e7-ba53-4189-b062-81fc64582431",
            "createdAt": "2020-11-13T12:00:00.123",
            "productId": "a269a247-0d38-4b47-9630-79c9ae545b68",
            "amountSold": 1,
            "name": "test"
          },
    ],
    isLoading: false,
    productList: null
}

export const articleComponentProps = {
    articles: [
        {
            id: "testID",
            name: "article name",
            amountInStock: 5,  
            amountToSubtract: 2,
            amountRequired: 4
        }
    ],
    updateArticle: () =>{},
    setArticles: () => {}
}

export const productDetailsProps = {
    "id": "a269a247-0d38-4b47-9630-79c9ae545b68",
    "name": "Dining Chair",
    "articles": [
        {
            "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
            "amountRequired": 4,
            "amountInStock": 2,
            "name": "article 1"
        },
        {
            "id": "831b92b8-677b-42cc-a585-335ea4ccccb6",
            "amountRequired": 5,
            "amountInStock": 1,
            "name": "article 2"
        },
    ],
    "requiredArticles": [
        {
          "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
          "amountRequired": 4
        },
    ],
    saleProduct : () =>{
        
    }
        
}

export const mockForValidate = {
    product: "test",
    requiredArticles: [{"id":"0517f083-0e15-4876-8d1f-6fa45900431c","amountRequired":12}],
    allArticles: [{"id":"0517f083-0e15-4876-8d1f-6fa45900431c","name":"Patched Leg","amountInStock":12},{"id":"addc65a8-c759-41d8-a18a-89fe446ad484","name":"Screw","amountInStock":41},{"id":"831b92b8-677b-42cc-a585-335ea4ccccb6","name":"Seat","amountInStock":8},{"id":"6892b98b-9b87-4520-9a9e-7528f1d78cb4","name":"Table Top","amountInStock":12},{"id":"41a84cdc-c70b-4924-9907-319572a45ca3","name":"Leg","amountInStock":12}]
}