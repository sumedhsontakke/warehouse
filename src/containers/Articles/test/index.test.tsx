
import { render, screen, waitFor } from "@testing-library/react";
import ArticlesContainer from "../index";
import {BrowserRouter as Router} from 'react-router-dom';
import * as request from "../../../misc/request";
import { act, } from 'react-dom/test-utils';
import { articleMockResponse } from '../../../mock';

describe('<ArticlesContainer />' , () =>{        
    it('should render and match the snapshot', async() => {
      const mock = jest.spyOn(request, "getAllArticles")
      .mockResolvedValue(articleMockResponse);
     
        const {
          container: { firstChild },
        } = 
        render(
            <Router>
                <ArticlesContainer />
            </Router>
        );
        await act(async () => {
          expect(firstChild).toMatchSnapshot();
          mock.mockRestore();
        });  
    }); 
    
    it('should render error modal and match shanpshot', async() => {
      const mock = jest.spyOn(request, "getAllArticles")
      .mockRejectedValue(null);
     
        const {
          container: { firstChild },
        } = 
        render(
            <Router>
                <ArticlesContainer />
            </Router>
        );
        act(() => {
          expect(firstChild).toMatchSnapshot();
          mock.mockRestore();
        });  
    });    
})