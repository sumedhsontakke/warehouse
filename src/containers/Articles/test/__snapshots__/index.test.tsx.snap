// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticlesContainer /> should render and match the snapshot 1`] = `
<div>
  <header
    class="sc-hKMtZM hMjKjB"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <img
            alt="logo"
            class="sc-dkzDqf ineiPS"
            src="sample-logo.png"
          />
        </div>
        <div
          class="col"
        >
          <nav
            class="sc-eCYdqJ fiquXb"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/sales"
            >
              Sales
            </a>
            <a
              href="/articles"
            >
              Articles
            </a>
          </nav>
        </div>
      </div>
    </div>
  </header>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="sc-jSMfEi hIJlZu"
        />
      </div>
    </div>
  </div>
  <div
    class="sc-iBkjds gwzIYY"
  >
    <div
      class="spinner-border text-primary"
      role="status"
    >
      <span
        class="visually-hidden"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`<ArticlesContainer /> should render error modal and match shanpshot 1`] = `
<div>
  <header
    class="sc-hKMtZM hMjKjB"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <img
            alt="logo"
            class="sc-dkzDqf ineiPS"
            src="sample-logo.png"
          />
        </div>
        <div
          class="col"
        >
          <nav
            class="sc-eCYdqJ fiquXb"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/sales"
            >
              Sales
            </a>
            <a
              href="/articles"
            >
              Articles
            </a>
          </nav>
        </div>
      </div>
    </div>
  </header>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="sc-jSMfEi hIJlZu"
        />
      </div>
    </div>
  </div>
  <div
    class="sc-iBkjds gwzIYY"
  >
    <div
      class="spinner-border text-primary"
      role="status"
    >
      <span
        class="visually-hidden"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
